R/phylo_check.R

```diff
@@ -0,0 +1,133 @@
+#' library(paleotree)
+#' data(RaiaCopesRule)
+#' plot(ceratopsianTreeRaia)
+#'
+#' #Specify list of names
+#' dinosaurs <- c("Nasutoceratops_titusi", "Diabloceratops_eatoni",
+#' "Zuniceratops_christopheri", "Psittacosaurus_major",
+#' "Psittacosaurus_sinensis", "Avaceratops_lammersi",
+#' "Xenoceratops_foremostensis", "Leptoceratops_gracilis",
+#' "Triceratops_horridus", "Triceratops_prorsus")
+#'
+#' #Counts of taxa in list, tree or both
+#' phylo_check(tree = ceratopsianTreeRaia, list = dinosaurs)
+#'
+#' #Table of taxon names in list, tree or both
+#' phylo_check(tree = ceratopsianTreeRaia, list = dinosaurs, out = "table")
+#'
+#' #Trim tree to tips in the list
+#' my_ceratopsians <- phylo_check(tree = ceratopsianTreeRaia, list = dinosaurs,
+#' out = "tree")
+#' plot(my_ceratopsians)
+#' @export
+
+phylo_check <- function(tree = NULL, list = NULL, out = "counts") {
+  #Errors for incorrect input
+  if (is.null(tree)) {
+    stop("Phylogeny must be provided")
+  }
+
+  if (inherits(tree, "phylo") == FALSE) {
+    stop("Phylogeny must be a phylo object")
+  }
+
+  if (is.null(list)) {
+    stop("List of taxa to check against must be provided")
+  }
+
+  if (is.vector(list) == FALSE) {
+    stop("List of taxa must be a vector")
+  }
+
+  if (any(grepl("[^[:alnum:][:space:]_]", list))) {
+    stop("List of taxa should not contain punctuation except spaces or
+         underscores")
+  }
+
+  if (out != "counts" && out != "table" && out != "tree") {
+    stop("out must either be counts, table or tree")
+  }
+
+  #Function
+  #Replace any spaces in taxon names with underscores
+  list <- gsub(" ", "_", list)
+  tree$tip.label <- gsub(" ", "_", tree$tip.label)
+
+  #Give uniform capitalisation
+  list <- gsub("^([a-z])", "\\U\\1", tolower(list), perl = TRUE)
+  tree$tip.label <- gsub("^([a-z])", "\\U\\1", tolower(tree$tip.label),
+                         perl = TRUE)
+
+  #Create vectors of names, those in tree and those in list
+  if (out == "counts" || out == "table") {
+    tip_names <- tree$tip.label
+    all_names <- unique(c(tip_names, list))
+    names_in_tree <- (all_names %in% tip_names)
+    names_in_list <- (all_names %in% list)
+  }
+
+  #Determine which names are in which lists and sum them up
+  if (out == "counts") {
+    tree_counts <- as.numeric(names_in_tree)
+    list_counts <- as.numeric(names_in_list)
+    difference <- tree_counts - list_counts
+
+    only_tree <- length(which(difference == 1))
+    only_list <- length(which(difference == -1))
+    in_both <- length(which(difference == 0))
+
+    counts <- data.frame(c("Tree and list", "Only in tree", "Only in list"),
+                        c(in_both, only_tree, only_list))
+    colnames(counts) <- c("Category", "Number of taxa")
+
+    return(counts)
+  }
+
+  #Determine which names are in which lists and table them
+  if (out == "table") {
+    table <- data.frame(all_names, names_in_tree, names_in_list)
+    table <- table[order(table$all_names), ]
```

Comment:
Part of me thinks the taxa in the list should be at the top of the table?

Reply:
I think an alphabetical sort would be helpful for identifying e.g. spelling mistakes, but agree that a presence sort would be most useful for seeing what is in the tree, so I've added this as a choice in the function. What do you think? Happy to change the default to 'presence' if you think that would be better.

Reply:
I personally think the 'presence' sort would be most useful, because I'm normally interested in which taxa in my list are not in my tree (or vice versa), but I suppose the next step is figure out why, and that could be because of spelling mistakes...I guess I'm really happy either way but 'presence' seems like the most immediately useful to me.

Reply:
Done :)

Reply:
About the "table" output format, maybe a table listing all taxa in both tree and list will be too long with very large trees/lists of taxa? Comparing this with the output from geiger's `name.check`, they only mention the taxa in the tree but not in the list ("Tree.not.data") and the opposite ("Data.not.tree"), so it will only inform you what's missing. Maybe we want both output formats? A more exhaustive one, as we currently have, a more concise one.